onnx-opl/src/non_max_suppression.rs

```diff
@@ -0,0 +1,277 @@
+            }
+            BoxRepr::CenterWidthHeight => {
+                let (box1_width_half, box1_height_half) = (box1[[2]] / 2.0, box1[[3]] / 2.0);
+                let (box2_width_half, box2_height_half) = (box2[[2]] / 2.0, box2[[3]] / 2.0);
+
+                let (x1_min, x1_max) = (box1[[0]] - box1_width_half, box1[[0]] + box1_width_half);
+                let (x2_min, x2_max) = (box2[[0]] - box2_width_half, box2[[0]] + box2_width_half);
+
+                let (y1_min, y1_max) = (box1[[1]] - box1_height_half, box1[[1]] + box1_height_half);
+                let (y2_min, y2_max) = (box2[[1]] - box2_height_half, box2[[1]] + box2_height_half);
+
+                (x1_min, x1_max, x2_min, x2_max, y1_min, y1_max, y2_min, y2_max)
+            }
+        };
+
+        let intersection_y_min = f32::max(y1_min, y2_min);
+        let intersection_y_max = f32::min(y1_max, y2_max);
+        if intersection_y_max <= intersection_y_min {
+            return false;
+        }
+
+        let intersection_x_min = f32::max(x1_min, x2_min);
+        let intersection_x_max = f32::min(x1_max, x2_max);
+        if intersection_x_max <= intersection_x_min {
+            return false;
+        }
+
+        let intersection_area =
+            (intersection_x_max - intersection_x_min) * (intersection_y_max - intersection_y_min);
+
+        if intersection_area <= 0.0 {
+            return false;
+        }
+
+        let area1 = (x1_max - x1_min) * (y1_max - y1_min);
+        let area2 = (x2_max - x2_min) * (y2_max - y2_min);
+
+        let union_area = area1 + area2 - intersection_area;
+
+        if area1 <= 0.0 || area2 <= 0.0 || union_area <= 0.0 {
+            return false;
+        }
+
+        let intersection_over_union = intersection_area / union_area;
+
+        intersection_over_union > iou_threshold
+    }
+}
+
+#[derive(Debug, Clone, Hash)]
+pub struct NonMaxSuppression {
+    pub center_point_box: BoxRepr,
+    pub num_selected_indices_symbol: Symbol,
+}
+
+impl_dyn_hash!(NonMaxSuppression);
+
+impl Op for NonMaxSuppression {
+    fn name(&self) -> Cow<str> {
+        "NonMaxSuppression".into()
+    }
+
+    fn op_families(&self) -> &'static [&'static str] {
+        &["onnx"]
+    }
+
+    op_as_typed_op!();
+}
+
+impl EvalOp for NonMaxSuppression {
+    fn is_stateless(&self) -> bool {
+        true
+    }
+
+    fn eval(&self, mut inputs: TVec<Arc<Tensor>>) -> TractResult<TVec<Arc<Tensor>>> {
+        let mut inputs = inputs.drain(..);
+        let boxes = inputs.next().ok_or(anyhow!("Expected at least 2 args"))?;
+        let scores = inputs.next().ok_or(anyhow!("Expected at least 2 args"))?;
+        let mut max_output_boxes_per_class =
+            inputs.next().map_or(Ok(0i64), |val| val.to_scalar::<i64>().copied())?;
+        let iou_threshold =
+            inputs.next().map_or(Ok(0f32), |val| val.to_scalar::<f32>().copied())?;
+        let score_threshold = inputs
```

Comment:
Ach. ok. maybe not.

Reply:
Mmm... Is this working when some inputs are skipped ? like when max_output_boxes_per_class is not there but iou_threashold is? Because in ONNX you can skip an input by using the empty string (ewww). tract does not allow such a thing :)

I usually proceed like here: have the optional input id as a field...
https://github.com/sonos/tract/blob/main/hir/src/ops/array/strided_slice.rs#L6

And then plug them in like this:
https://github.com/sonos/tract/blob/main/onnx/src/ops/array/slice.rs#L129

Reply:
All right :) I just saw the Expansion and... you actually found about that, well done... But how does the -opl eval() know which optional input is which ? Unless I'm mistaken, we're loosing this information at Expansion::wire() time.

Leaving you with a couple options, I think:
* either propagate the optional_*_input stuff in the -opl op
* decide that these optional inputs have to be constant, and turn them into value fields in the -opl op
* decide that they will be always here: we can wire a const (target.add_const in Expansion::wire) with the default value when they're missing

Please consider option #2 seriously: it may be less satisfactory as it is less generic, but one of my strategies to avoid tract being bloating with useless code is to wait for evidence of usage for stuff that looks a bit exotic... (arguably this lead to half baked ops like Resize...).

Reply:
I prefer the third approach, it's probably the cleanest and changing the current code to use it shouldn't add more than one or two lines of code. If I understand correctly, it's also how the nnef code does optional arguments, right?

I can do the second approach if you want to though.

Reply:
Go for the third approach, no problem.

Optional args in NNEF ? they are handled by the syntax (as it has named parameters). Then it's a mix of accepting tensors as inputs or coercing to single integer values depending on the operators, I would say.

Reply:
@kali I did the changes, let me know if they are ok or if you want anything else changed :)